File: duna-core/src/main/java/io/duna/core/ContextImpl.java
package io.duna.core;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class ContextImpl implements Context {

    private static final ThreadLocal<Context> contextHolder = new ThreadLocal<>();

    private Duna manager;
    private Map<String, Object> data;

    public ContextImpl(Duna manager) {
        this.manager = manager;
        this.data = new ConcurrentHashMap<>();
    }

    static Context currentContext() {
        return contextHolder.get();
    }

    public static void setContext(Context context) {
        contextHolder.set(context);
    }

    @Override
    public Duna manager() {
        return this.manager;
    }

    @Override
    public <T> T get(String key) {
        return (T) data.get(key);
    }

    @Override
    public <T> void put(String key, T value) {
        data.put(key, value);
    }

    @Override
    public <T> void putAll(Map<String, T> data) {
        this.data.putAll(data);
    }

    @Override
    public void remove(String key) {
        data.remove(key);
    }
}

File: duna-core/src/main/kotlin/io/duna/core/ContextImpl.kt
package io.duna.core

import java.util.concurrent.ExecutorService

internal class ContextImpl(private val executorService: ExecutorService,
                           private val workerPool: ExecutorService)
  : Context, AbstractMap<String, Any>() {

  private val internalMap = HashMap<String, Any>()

  override val entries: MutableSet<MutableMap.MutableEntry<String, Any>>
    get() = internalMap.entries

  override val keys: MutableSet<String>
    get() = internalMap.keys

  override val values: MutableCollection<Any>
    get() = internalMap.values

  override fun execute(block: () -> Unit) {
    if (Context.inDunaThread)
      block.invoke()
    else
      executorService.submit(block)
  }

  override fun executeBlocking(block: () -> Unit) {
    workerPool.submit(block)
  }

  override fun containsKey(key: String): Boolean {
    return super.containsKey(key)
  }

  override fun clear() {
    internalMap.clear()
  }

  override fun get(key: String): Any? {
    return internalMap[key]
  }

  override fun put(key: String, value: Any): Any? {
    return internalMap.put(key, value)
  }

  override fun putAll(from: Map<out String, Any>) {
    internalMap.putAll(from)
  }

  override fun remove(key: String): Any? {
    return internalMap.remove(key)
  }
}
